// stage/build/cli/util/viteConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.viteConfig = void 0;
const plugin_react_1 = __importDefault(require("@vitejs/plugin-react"));
const plugin_node_resolve_1 = __importDefault(require("@rollup/plugin-node-resolve"));
const plugin_json_1 = __importDefault(require("@rollup/plugin-json"));
const path_1 = __importDefault(require("path"));
const appRoot = path_1.default.resolve(process.cwd(), "app");
const stagePath = path_1.default.resolve(__dirname, "..", "..", "..", "frontend");
process.env.STAGE_APP_PATH = path_1.default.resolve(process.cwd());
exports.viteConfig = {
    plugins: [(0, plugin_react_1.default)()],
    root: stagePath,
    resolve: {
        alias: [
            { find: "@app", replacement: appRoot },
            { find: "stage/client", replacement: path_1.default.resolve(stagePath, "lib") },
        ],
    },
    build: {
        rollupOptions: {
            plugins: [
                (0, plugin_node_resolve_1.default)({
                    browser: true,
                    preferBuiltins: false,
                    dedupe: ["react", "@reach/router"],
                }),
                (0, plugin_json_1.default)(),
            ],
        },
    },
};
